Psgs integration

daehiff opened the pull request.

daehiff committed:
Basic setup

daehiff committed:
Integrate utility for paulipolynomials

daehiff committed:
Add logic representing pauli polynomials

daehiff committed:
Cleanup of toplevel files

daehiff committed:
Create basic infrastructure for synthesis algorithms

daehiff committed:
Test basic infrastructure

daehiff committed:
Add in most relevant synthesis algorithms

daehiff committed:
Smaller fixes

daehiff committed:
Adds in X, Y and Z to preprend of a tableau

Signed-off-by: David Winderl <winderl13@gmail.com>

daehiff committed:
Note the clifford application on gates rather than on a tableau

daehiff committed:
Use propagation rules for propagating a pauli into a clifford

daehiff committed:
Reformat pauliopt

y-richie-y commented on pauliopt/clifford/tableau.py:
```diff
@@ -37,24 +61,24 @@ def mult_paulis(p1, p2, sign1, sign2, n_qubits):
     """
-    Class for storing and manipulating Clifford tableau.
-    A Clifford tableau is a representation of a Clifford circuit as a
+    Class for storing and manipulating Clifford clifford.
```
I think there are search and replace problems across the PR about clifford <-> tableau, please fix them

daehiff replied:
Ohh boi, thx! :D

y-richie-y commented on pauliopt/circuits.py:
```diff
@@ -92,6 +100,36 @@ def _check_gate(self, gate):
+        inverted.add_gates(inv_gates)
+        return inverted
+
+    def apply_permutation(self, permutation: list):
```
`list[int]`?

y-richie-y commented on pauliopt/gates.py:
```diff
@@ -112,18 +116,158 @@ def gadgets(self):
         return f"{self.name}({self.phase}, {', '.join(args)})"
 
+    def get_phase_as_float(self):
+        return self.phase if not isinstance(self.phase, Angle) else self.phase.to_qiskit
```
so weird that this is `to_qiskit`, does `float(.)` not work?

daehiff replied:
That is the implementation of `to_qiskit`:

```
    @property
    def to_qiskit(self) -> float:
        return float(self)
```

The whole angle type object is kind of a difficult and un-understandable interface, might be sensible to also refactor it or remove the type completely.

I.e.: I will change it :D

y-richie-y commented on pauliopt/gates.py:
```diff
@@ -112,18 +116,158 @@ def gadgets(self):
+    @abstractmethod
+    def get_h_s_cx_decomposition(self) -> List[Union["H", "S", "CX"]]:
+        """
+        Every clifford must me decomposable into a list of H, S and CX gates.
```
must be

y-richie-y commented on pauliopt/gates.py:
```diff
@@ -112,18 +116,158 @@ def gadgets(self):
+        """
+        h_s_cx_decomposition = self.get_h_s_cx_decomposition()
+        for gate in reversed(h_s_cx_decomposition):
+            if gate.name == "H":
```
the logic of these 3 cases seem quite similar. Is it possible to combine it into one by having a mapping from gate.name to propagation dictionary?

y-richie-y replied:
and a loop to update the Paulis?

daehiff replied:
What exactly do you mean here loop over the qubits? 
But of a CX a control is kind of dependent of the target pauli and vice versa.

y-richie-y commented on pauliopt/gates.py:
```diff
@@ -421,15 +695,17 @@ class Rx(PhaseGate):
     def decomp(self):
         (q,) = self.qubits
-        return [XHead(self.phase) @ {q}]
+        if isinstance(self.phase, Angle):
+            return [XHead(self.get_phase_as_angle()) @ {q}]
+        else:
+            return [XHead(Angle(self.get_phase_as_angle())) @ {q}]
```
Why can't we just call `get_phase_as_angle`? This is what it does

y-richie-y commented on pauliopt/pauli/pauli_gadget.py:
```diff
@@ -56,101 +59,180 @@ def find_minimal_cx_assignment(column: np.array, arch: Topology):
 
 class PauliGadget:
-    def __init__(self, angle: AngleExpr, paulis: List[Pauli]):
+    def __init__(self, angle: float, paulis: List[Pauli]):
```
Why do we need to change the type here?

y-richie-y commented on pauliopt/pauli/pauli_polynomial.py:
```diff
@@ -1,41 +1,153 @@
+            legs += gadget.num_legs()
+        return legs
+
+    def append_pauli_gadget(self, pauli_gadget: PauliGadget):
```
Do we need to have this method?

y-richie-y commented on pauliopt/pauli/pauli_polynomial.py:
```diff
@@ -94,7 +239,7 @@ def to_svg(
         text_width = int(math.ceil(50 * vscale))
 
-        bend_degree = 10
+        bend_degree = int(math.ceil(10))
```
just 10

daehiff committed:
Fix PR

Signed-off-by: David Winderl <winderl13@gmail.com>

Removes append_pauli gadegt

must me -> must be

Change type to float constructor

Signed-off-by: David Winderl <winderl13@gmail.com>

Clifford clifford -> Clifford Tableau

Signed-off-by: David Winderl <winderl13@gmail.com>

Removes redundant if condition

Signed-off-by: David Winderl <winderl13@gmail.com>

Fix unit tests and pipeline

Signed-off-by: David Winderl <winderl13@gmail.com>

Fix tests

Signed-off-by: David Winderl <winderl13@gmail.com>

daehiff committed:
Remove python 3.9 and add in 3.10, 3.11 and 3.12

Signed-off-by: David Winderl <winderl13@gmail.com>

y-richie-y committed:
Small change

y-richie-y commented on pauliopt/pauli/pauli_polynomial.py:
```diff
@@ -1,41 +1,150 @@
+        return legs
+
+    def assign_time(self, time: float):
+        for gadet in self.pauli_gadgets:
```
gadget

y-richie-y commented on pauliopt/pauli/pauli_polynomial.py:
```diff
@@ -1,41 +1,150 @@
+        else:
+            pad_len = 0
+        return "\n".join(
+            [self[i].to_string(pad_lenght=pad_len) for i in range(self.num_gadgets)]
```
pad_length

y-richie-y commented on pauliopt/pauli/pauli_gadget.py:
```diff
@@ -74,83 +81,159 @@ def __len__(self):
+    def num_qubits(self):
+        return len(self.paulis)
+
+    def to_string(self, pad_lenght=0):
```
length!

y-richie-y commented on pauliopt/pauli/synthesis/synthesis_divide_and_conquer.py:
```diff
@@ -0,0 +1,220 @@
+    )
+
+
+def sort_pauli_polynomial(pp: PauliPolynomial) -> PauliPolynomial:
```
Would the algorithm be faster if we used a different kind of sort?
Could we take advantage of the builtin python sort function here?

daehiff replied:
Yes, but that would require a lot of refactoring since a also checks commutation and I am not confident of doing that now.

y-richie-y replied:
Yes lets do this in the future instead :)

y-richie-y commented on pauliopt/pauli/synthesis/steiner_gray_synthesis.py:
```diff
@@ -0,0 +1,536 @@
+
+    cols.sort(key=lambda x: len(x))
+
+    return cols_i, cols[-1][1:], cols[-1][0], cols[0][1:] + cols[1][1:]
```
Maybe cols should be a dictionary instead

daehiff replied:
Same argument as above, the point was to first integrate our algorithms

daehiff committed:
Fix PR
